File: 01._Assignments/01a._Data_parsing_server_Part_I/node_data_parsing_server/dataParser.js
const fs = require('node:fs');
const xml2js = require('xml2js');
const yaml = require('js-yaml');
const fastcsv = require('fast-csv')

function parseDataFile(path){

    let filePathArray = path.split('.');
    let filenameExtension = filePathArray[filePathArray.length-1];

    fs.readFile(path, 'utf8', (err, data) => {
        if (err) {
          console.error('Error reading file:', err);
          return;
        }

    switch(filenameExtension){
        case 'txt':{
            console.log(data)
            break;
        }
        case 'xml':{
            parser = new xml2js.Parser()
            parser.parseString(data, (err, result)=>{ // vi bruger et callback til at håndtere resultatet når metoden har parset
                if(err){
                    console.error(err)
                }else{
                    console.dir(result, {depth: null}); //se xml2js docs - Node uses util.inspect to convert the object into strings and that function stops after depth=2 which is a bit low for most XML.
                }
            })
            break;
        }
        case 'json':{
            const jsonData = JSON.parse(data)
            console.log(jsonData)
            console.log(jsonData.people[0].name)
            break;
        }
        case 'yaml':{
            const yamlData = yaml.load(data);
            console.log(yamlData);
            break;
        }
        case 'csv':{
            fastcsv.parseString(data, { headers: true }) 
            .on('data', (row) => console.log('CSV row:', row)); //read up
            break;
        }
        default:{
            console.log("file type not supported ;-)")
        }
    }
});
}
parseDataFile('./data/data.csv')

File: 01._Assignments/01a._Data_parsing_server_Part_I/python_data_parsing_server/dataParser.py
import os
import json
import yaml
import csv
import xml.etree.ElementTree as ET

def parse_data_file(path):
    _, file_extension = os.path.splitext(path)
    file_extension = file_extension.lower()[1:]  # Remove the leading dot

    try:
        with open(path, 'r', encoding='utf-8') as file:
            data = file.read()
        
        if file_extension == 'txt':
            print(data)

        elif file_extension == 'xml':
            root = ET.fromstring(data)
            print(ET.tostring(root, encoding='utf-8').decode())

        elif file_extension == 'json':
            json_data = json.loads(data)
            print(json_data)
            print(json_data.get('people', [{}])[0].get('name', 'No name found'))

        elif file_extension in ['yaml', 'yml']:
            yaml_data = yaml.safe_load(data)
            print(yaml_data)

        elif file_extension == 'csv':
            reader = csv.DictReader(data.splitlines())
            for row in reader:
                print("CSV row:", row)

        else:
            print("File type not supported ;-)")

    except Exception as e:
        print(f"Error reading file: {e}")

parse_data_file('./data/data.yaml')
